File: task-manager-ios/task-manager-ios/Presentation/Core/TasksListViewModel.swift
//
//  TasksListViewModel.swift
//  task-manager-ios
//
//  Created by Jeofferson Dela Peña on 3/9/24.
//

import Foundation

class TasksListViewModel: ObservableObject {
    @Published var state = TasksListState()
    
    private let repository: TaskRepository
    
    init(repository: TaskRepository) {
        self.repository = repository
    }
    
    func getTasks(isFromSwipeRefresh: Bool = false) {
        if !isFromSwipeRefresh {
            state.isLoading = true
        }
        
        switch repository.getTasks() {
        case .success(let tasks):
            state.tasks = tasks
        case .failure(let error):
            state.activeAlert = .constant(.error(error))
        }
        
        if !isFromSwipeRefresh {
            state.isLoading = false
        }
    }
    
    func addTask(_ task: Task) {
        state.isLoading = true
        
        switch repository.addTask(task) {
        case .success:
            state.activeSheet = .constant(nil)
            getTasks()
        case .failure(let error):
            state.activeAlert = .constant(.error(error))
        }
        
        state.isLoading = false
    }
    
    func editTask(_ task: Task) {
        state.isLoading = true
        
        switch repository.editTask(task) {
        case .success:
            state.activeSheet = .constant(nil)
            getTasks()
        case .failure(let error):
            state.activeAlert = .constant(.error(error))
        }
        
        state.isLoading = false
    }
    
    func deleteTask(_ task: Task) {
        state.isLoading = true
        
        switch repository.deleteTask(task) {
        case .success:
            state.activeAlert = .constant(nil)
            state.activeSheet = .constant(nil)
            getTasks()
        case .failure(let error):
            state.activeAlert = .constant(.error(error))
        }
        
        state.isLoading = false
    }
    
    static let sample = TasksListViewModel(
        repository: TaskRepositoryImpl(
            remoteService: TaskRemoteService(),
            localService: TaskLocalService()
        )
    )
}

File: task-manager-android/app/src/main/java/com/jeoffersondelapena/task_manager_android/presentation/core/TasksListViewModel.kt
package com.jeoffersondelapena.task_manager_android.presentation.core

import android.content.Context
import androidx.lifecycle.ViewModel
import com.jeoffersondelapena.task_manager_android.data.local.service.TaskLocalService
import com.jeoffersondelapena.task_manager_android.data.remote.service.TaskRemoteService
import com.jeoffersondelapena.task_manager_android.data.repository.TaskRepositoryImpl
import com.jeoffersondelapena.task_manager_android.domain.model.Task
import com.jeoffersondelapena.task_manager_android.domain.repository.TaskRepository
import com.jeoffersondelapena.task_manager_android.domain.util.helper.TaskManagerResult
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject

@HiltViewModel
class TasksListViewModel @Inject constructor(
    private val repository: TaskRepository
) : ViewModel() {
    val state = TasksListState

    fun getTasks(isFromSwipeRefresh: Boolean = false) {
        if (!isFromSwipeRefresh) {
            state.isLoading = true
        }

        when (val result = repository.getTasks()) {
            is TaskManagerResult.Success -> {
                state.tasks.value = result.value
            }
            is TaskManagerResult.Failure -> {
                state.activeDialog.value = TasksListState.DialogType.Error(result.exception)
            }
        }

        if (!isFromSwipeRefresh) {
            state.isLoading = false
        }
    }

    fun addTask(task: Task) {
        state.isLoading = true

        when (val result = repository.addTask(task)) {
            is TaskManagerResult.Success -> {
                state.activeModalBottomSheet.value = null
                getTasks()
            }
            is TaskManagerResult.Failure -> {
                state.activeDialog.value = TasksListState.DialogType.Error(result.exception)
            }
        }

        state.isLoading = false
    }

    fun editTask(task: Task) {
        state.isLoading = true

        when (val result = repository.editTask(task)) {
            is TaskManagerResult.Success -> {
                state.activeModalBottomSheet.value = null
                getTasks()
            }
            is TaskManagerResult.Failure -> {
                state.activeDialog.value = TasksListState.DialogType.Error(result.exception)
            }
        }

        state.isLoading = false
    }

    fun toggleTaskCompletion(task: Task) {
        state.isLoading = true

        when (val result = repository.toggleTaskCompletion(task)) {
            is TaskManagerResult.Success -> {
                state.activeModalBottomSheet.value = null
                getTasks()
            }
            is TaskManagerResult.Failure -> {
                state.activeDialog.value = TasksListState.DialogType.Error(result.exception)
            }
        }

        state.isLoading = false
    }

    fun deleteTask(task: Task) {
        state.isLoading = true

        when (val result = repository.deleteTask(task)) {
            is TaskManagerResult.Success -> {
                state.activeDialog.value = null
                state.activeModalBottomSheet.value = null
                getTasks()
            }
            is TaskManagerResult.Failure -> {
                state.activeDialog.value = TasksListState.DialogType.Error(result.exception)
            }
        }

        state.isLoading = false
    }

    companion object {
        fun sample(context: Context): TasksListViewModel {
            return TasksListViewModel(
                TaskRepositoryImpl(
                    TaskRemoteService(),
                    TaskLocalService(context),
                )
            )
        }
    }
}
